internal/driver/driver.go

```diff
@@ -87,113 +85,132 @@ func (s *Driver) Initialize(sdk interfaces.DeviceServiceSDK) error {
 	var batch_size = 16
+
 	var reqs_len = len(reqs)
-	var items = []gos7.S7DataItem{}
-	res = make([]*sdkModel.CommandValue, reqs_len)
+	var s7DataItems = []gos7.S7DataItem{}
 
-	var errors = make([]string, reqs_len)
+	res = make([]*sdkModel.CommandValue, reqs_len)
+	var s7_errors = make([]string, reqs_len)
 
-	// Init a two-dimensional array
-	var datas = make([][]byte, reqs_len)
-	for i := range datas {
-		datas[i] = make([]byte, 4) // 4 bytes
+	// two-dimensional array for handle S7DataItems
+	var dataset = make([][]byte, reqs_len)
+	for i := range dataset {
+		dataset[i] = make([]byte, 4) // 4 bytes
 	}
 
 	// Get S7 device connection information, each Device has its own connection.
-	s7Client := s.getS7Client(deviceName)
-	if s7Client == nil {
-		s.lc.Errorf("Can not get S7CLient from: %s", deviceName)
-		return
-	}
+	s7Client := s.getS7Client(deviceName, protocols)
 
-	// assembly items
-	times := int(reqs_len / batch_size)
-	remains := reqs_len % batch_size
+	// assemble s7DataItems, get items, fetch data
+
+	var times = int(reqs_len / batch_size)
+	var remains = reqs_len % batch_size
+	var tmp_reqs []sdkModel.CommandRequest
 
-	var reqs1 []sdkModel.CommandRequest
 outloop:
 	for j := 0; j <= times; j++ {
 
 		// 1. init array
-		items = items[:0] // clear array
-		reqs1 = reqs1[:0] // clear array
+		s7DataItems = s7DataItems[:0] // clear array
+		tmp_reqs = tmp_reqs[:0]       // clear array
 		if j >= times {
 			if remains > 0 {
-				reqs1 = reqs[times*batch_size : times*batch_size+remains]
+				tmp_reqs = reqs[times*batch_size : times*batch_size+remains]
 			} else {
 				break outloop // load complete, exit loop
 			}
 		} else {
-			reqs1 = reqs[j*batch_size : j*batch_size+batch_size]
+			tmp_reqs = reqs[j*batch_size : j*batch_size+batch_size]
 		}
 
 		// 2. get resources from reqs append to items
-		for i, req := range reqs1 {
+		for i, req := range tmp_reqs {
 
-			variable := cast.ToString(req.Attributes["NodeName"])
-			dbInfo, _ := s.getDBInfo(variable)
-			// fmt.Println(dbInfo, err)
+			nodename := cast.ToString(req.Attributes["NodeName"])
+			dbInfo, _ := s.getDBInfo(nodename)
 
-			var item = gos7.S7DataItem{
+			var s7DataItem = gos7.S7DataItem{
 				Area:     dbInfo.Area,
 				WordLen:  dbInfo.WordLength,
 				DBNumber: dbInfo.DBNumber,
 				Start:    dbInfo.Start,
 				Amount:   dbInfo.Amount,
-				Data:     datas[j*batch_size+i],
-				Error:    errors[i],
+				Data:     dataset[j*batch_size+i],
+				Error:    s7_errors[i],
 			}
-
-			items = append(items, item)
+			s7DataItems = append(s7DataItems, s7DataItem)
 		}
+		s.lc.Debugf("Read from S7DataItems: ", s7DataItems)
+
+		// 3. use AGReadMulti api to get values from S7 device, if error, try 3 times
+		retrytimes := 3
+		for {
+			err = s7Client.Client.AGReadMulti(s7DataItems, len(s7DataItems))
+			if err != nil {
+				s.lc.Errorf("AGReadMulti Error: %s, reconnecting...", err)
+				s.mu.Lock()
+				s.s7Clients[deviceName] = nil
```

Comment:
Yes, it's right.

Reply:
why should this be unlock before `s.s7Clients[deviceName] = s.NewS7Client(deviceName, protocols)` ?
Since the code is still accessing the map, the Unlock should be called after it.

Aside from that, why do you need this line `s.s7Clients[deviceName] = nil`, since you are going to assign a value for this key?